File: src/modules/trees/1_representation/ArrayRepresentation.ts
// Mapping Binary tree to array

/**
 * To represent the binary tree in an array, we follow a specific mapping
 * 1) the root node is stored at index 0 in the array
 * 2) for any node at index i, it left child is stored at index 2*i+1, and the right child is stored at index 2*i+2
 * 3) if the node is not present, the array stores a special value (eg null,-1, or any senitel value)
 */

/**
 * example 
 *      10
       /  \
      5    20
     / \   / \
    3   8 15  25

    [10, 5, 20, 3, 8, 15, 25]

 */


class ArrayBinaryTree {
    private arr: number[]
    constructor(values: number[]) {
        this.arr = values
    }
    public getRoot(): number {
        return this.arr[0]
    }
    public getLeftChild(index: number): number {
        let leftIndex: number = 2 * index + 1
        if (leftIndex >= this.arr.length || this.arr[leftIndex] === null) return -1
        return this.arr[leftIndex]
    }
    public getRightChild(index: number): number {
        let rightIndex: number = 2 * index + 2
        if (rightIndex >= this.arr.length || this.arr[rightIndex] === null) return -1
        return this.arr[rightIndex]
    }
}

const values = [10, 5, 20, 3, 8, 15, 25];
const tree = new ArrayBinaryTree(values);

console.log("Root:", tree.getRoot()); // Output: Root: 10
console.log("Left child of 5:", tree.getLeftChild(1)); // Output: Left child of 5: 3
console.log("Right child of 5:", tree.getRightChild(1)); // Output: Right child of 5: 8

export{}

File: src/modules/trees/1_representation/ArrayRepresentation.go
package main

import "fmt"

type ArrayBinaryTree struct {
	arr []int
}

func NewArrayBinaryTree(values []int) *ArrayBinaryTree {
	return &ArrayBinaryTree{arr: values}
}

func (tree *ArrayBinaryTree) GetRoot() int {
	if len(tree.arr) > 0 {
		return tree.arr[0]
	}
	return -1
}

func (tree *ArrayBinaryTree) GetLeftChild(index int) int {
	leftIndex := 2*index + 1
	if leftIndex >= len(tree.arr) || tree.arr[leftIndex] == 0 {
		return -1
	}

	return tree.arr[leftIndex]
}

func (tree *ArrayBinaryTree) GetRightChild(index int) int {
	rightIndex := 2*index + 2
	if rightIndex >= len(tree.arr) || tree.arr[rightIndex] == 0 {
		return -1
	}

	return tree.arr[rightIndex]
}

func main() {
	values := []int{10, 5, 20, 3, 8, 15, 25}
	tree := NewArrayBinaryTree(values)

	fmt.Println("Root:", tree.GetRoot())                    // Output: Root: 10
	fmt.Println("Left child of 5:", tree.GetLeftChild(1))   // Output: Left child of 5: 3
	fmt.Println("Right child of 5:", tree.GetRightChild(1)) // Output: Right child of 5: 8
}
